pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -296,97 +295,89 @@ public void skipEntriesFailed(ManagedLedgerException exception, Object ctx) {
-                return future;
-            }
-        }
         PersistentMessageFinder persistentMessageFinder = new PersistentMessageFinder(topicName, cursor);
-        class Result {
-            Position position = null;
-        }
-
-        final Result result = new Result();
-        final CountDownLatch counter = new CountDownLatch(1);
 
         if (log.isDebugEnabled()) {
             log.debug("[{}][{}] Resetting subscription to timestamp {}", topicName, subName, timestamp);
         }
         persistentMessageFinder.findMessages(timestamp, new AsyncCallbacks.FindEntryCallback() {
             @Override
             public void findEntryComplete(Position position, Object ctx) {
-                result.position = position;
-                counter.countDown();
-            }
-
-            @Override
-            public void findEntryFailed(ManagedLedgerException exception, Object ctx) {
-                // todo - what can go wrong here that needs to be retried?
-                if (exception instanceof ConcurrentFindCursorPositionException) {
-                    future.completeExceptionally(new SubscriptionBusyException(exception.getMessage()));
+                final Position finalPosition;
+                if (position == null) {
+                    // this should not happen ideally unless a reset is requested for a time
+                    // that spans beyond the retention limits (time/size)
+                    finalPosition = cursor.getFirstPosition();
+                    if (finalPosition == null) {
+                        log.warn("[{}][{}] Unable to find position for timestamp {}. Unable to reset cursor to first position",
+                                topicName, subName, timestamp);
+                        future.completeExceptionally(
+                                new SubscriptionInvalidCursorPosition("Unable to find position for specified timestamp"));
+                        return;
+                    }
+                    log.info(
+                            "[{}][{}] Unable to find position for timestamp {}. Resetting cursor to first position {} in ledger",
+                            topicName, subName, timestamp, finalPosition);
                 } else {
-                    future.completeExceptionally(new BrokerServiceException(exception));
+                    finalPosition = position;
                 }
-                counter.countDown();
-            }
-        });
 
-        try {
-            counter.await();
-        } catch (Exception e) {
-            log.warn("[{}][{}] message find interrupted ", topicName, subName, e);
-            if (future.isCompletedExceptionally()) {
-                return future;
-            }
-            future.completeExceptionally(e);
-            return future;
-        }
+                if (!isFenced.compareAndSet(false, true)) {
+                    future.completeExceptionally(new SubscriptionBusyException("Failed to fence subscription"));
```

Comment:
I think we should be fine as we disconnect the consumers and reconnection of the consumer will take care of it.